eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelay.scala

```diff
@@ -0,0 +1,347 @@
+        case None =>
+          // TODO: @pm: maybe those checks should be done later in the flow (by the mpp FSM?)
+          context.log.warn("rejecting htlcId={}: missing payment secret", add.id)
+          rejectHtlc(add.id, add.channelId, add.amountMsat)
+          Behaviors.stopped
+        case Some(secret) =>
+          import akka.actor.typed.scaladsl.adapter._
+          context.log.debug("creating a payment FSM")
+          val mppFsm = context.actorOf(MultiPartPaymentFSM.props(nodeParams, add.paymentHash, outer.totalAmount, adapters.toClassic))
+          context.log.debug("forwarding incoming htlc to the payment FSM")
+          mppFsm ! MultiPartPaymentFSM.HtlcPart(outer.totalAmount, add)
+          receiving(Queue(add), secret, inner, next, mppFsm)
+        case _ => Behaviors.unhandled
+      }
+    }
+
+  /**
+   * We start by aggregating an incoming HTLC set. Once we received the whole set, we will compute a route to the next
+   * trampoline node and forward the payment.
+   *
+   * @param htlcs       received incoming HTLCs for this set.
+   * @param secret      all incoming HTLCs in this set must have the same secret to protect against probing / fee theft.
+   * @param nextPayload relay instructions (should be identical across HTLCs in this set).
+   * @param nextPacket  trampoline onion to relay to the next trampoline node.
+   * @param handler     actor handling the aggregation of the incoming HTLC set.
+   */
+  private def receiving(htlcs: Queue[UpdateAddHtlc], secret: ByteVector32, nextPayload: Onion.NodeRelayPayload, nextPacket: OnionRoutingPacket, handler: ActorRef): Behavior[Command] =
+    Behaviors.receiveMessage {
+      case WrappedNodeRelayPacket(IncomingPacket.NodeRelayPacket(add, outer, _, _)) => outer.paymentSecret match {
+        case None =>
+          context.log.warn("rejecting htlcId={}: missing payment secret", add.id)
+          rejectHtlc(add.id, add.channelId, add.amountMsat)
+          Behaviors.same
+        case Some(incomingSecret) if incomingSecret != secret =>
+          context.log.warn("rejecting htlcId={}: payment secret doesn't match other HTLCs in the set", add.id)
+          rejectHtlc(add.id, add.channelId, add.amountMsat)
+          Behaviors.same
+        case Some(incomingSecret) if incomingSecret == secret =>
+          context.log.debug("forwarding incoming htlc to the payment FSM")
+          handler ! MultiPartPaymentFSM.HtlcPart(outer.totalAmount, add)
+          receiving(htlcs :+ add, secret, nextPayload, nextPacket, handler)
+      }
+      case WrappedMultiPartPaymentFailed(MultiPartPaymentFSM.MultiPartPaymentFailed(_, failure, parts)) =>
+        context.log.warn("could not relay payment (paidAmount={} failure={})", parts.map(_.amount).sum, failure)
+        Metrics.recordPaymentRelayFailed(failure.getClass.getSimpleName, Tags.RelayType.Trampoline)
+        handler ! PoisonPill
```

Comment:
We need to do some experimentation, but now that you spawn a `NodeRelay` per-trampoline payment, we shouldn't need to explicitly `PoisonPill` the `MultiPartPaymentFSM`, it should be stopped automatically when the `NodeRelay` stops (which may be a bit later than before when we actually relay downstream, but shouldn't be an issue).